core-rust/state-manager/src/store/proofs_gc.rs

```diff
@@ -175,60 +181,346 @@ impl LedgerProofsGc {
         );
         let mut last_pruned_state_version = progress_started_at.epoch_proof_state_version;
 
+        let mut pruned_proof_ranges = vec![];
         let mut retained_proofs = 0; // only for logging purposes
         loop {
-            let batch_start_state_version_inclusive = last_pruned_state_version
-                .next()
-                .expect("state version overflow");
-
-            // Locate the next proof that we need to retain. We use the same method and limits as
-            // the business logic responsible for outputting proven transactions:
-            let locate_proof_database = self.database.read_current();
-            let transactions_and_proof = locate_proof_database.get_txns_and_proof(
-                batch_start_state_version_inclusive,
-                self.limits_config
-                    .max_txns_for_responses_spanning_more_than_one_proof,
-                self.limits_config.max_txn_bytes_for_single_response,
-            );
-            drop(locate_proof_database);
-
-            let Some((_transactions, proof)) = transactions_and_proof else {
-                error!(
-                    "A chain of transactions-without-proof from state version {} does not fit within the limits {:?}; aborting the GC",
-                    batch_start_state_version_inclusive, self.limits_config
-                );
-                return;
+            let Some(next_prune_range) = self.locate_next_prune_range(last_pruned_state_version)
+            else {
+                break;
```

Comment:
is it an opportunity to use that cool Rust idiom `while let Some(x) = potentially_get_next_x() {` ?